### PR TITLE
Line joiners: scanline and lookback

### PR DESCRIPTION
**Pull request by bugadani**
This PR reworks how the triangles in a polyline are rendered. Instead of trying to get the mathematical point rendering to work, I've instead created a fast-ish scanline iterator to draw individual triangles, and introduced a check in ThickPoints, that, for each to-be-rendered pixel checks, if it was part of a previous triangle. This check is done for the last 3 triangles to avoid overdraw on the inside of bevelled corners.

# The scanline iterator

The new scanline iterator (`FillScanlineIterator` for now) returns the points of a triangle from top to bottom, from left to right. This implementation differs from the existing `ScanlineIterator` in that it does not differentiate between edge and inside points, and also only yields unique points, thus it has no overdraw.

Basically, it walk the outside edges, determines which are the limits of a line and for each line, creates a line point iterator to generate the points.

Maybe some of it's implementation can be used to improve the `ScanlineIterator` but I haven't bothered as it was tricky enough to get this to work.

This PR replaces the implementation behind Triangle::points() to use the new `FillScanlineIterator`.

# Lookback

Right now, the lookback in `ThickPoints` contains 3 Option<Triangle> fields and for each of those, checks if the point we want to return is inside any of the 3 previous triangles. This isn't particularly efficient. It should be possible to only store the unique points and reconstruct triangles, so the memory cost can be cut down from 9 points + extra to 5 points + a little less, but the runtime cost can't be reduced this way.

There are some development options:
 - Try to solve the overdraw by tuning the bevel rendering - a lookback of 1 triangle is still necessary to avoid drawing over the
   triangle edges.
 - Try to solve edge overdraw some other way. Maybe it is possible to only track vertical information of the previous triangle and 
   abuse the fact that the scanline iterator returns points in a given order.

---------------

This PR does NOT solve the general overdraw issue.

This PR is opened agains #386 so the code is not master-quality

### CODE REVIEW

**Comment by bugadani** [comment]
There is still this one:
image

But the overdraw issue in the original PR (https://user-images.githubusercontent.com/800791/90419139-0883ad00-e0ae-11ea-9216-e00955c897ff.png) is gone

File: embedded-graphics/src/primitives/line_joint.rs
Comment (bugadani) [inline embedded-graphics/src/primitives/line_joint.rs:205]: I have no clue at all why this helps overdraw.

**Comment by bugadani** [comment]
@jamwaffles for some reason I can't check CI but I believe the failures are the same as with the parent PR.

File: embedded-graphics/src/primitives/line_joint.rs
Comment (jamwaffles) [reply embedded-graphics/src/primitives/line_joint.rs:205]: It changes the winding order to be clockwise which I think has some effect on the point-in-triangle algorithm. There must be some numerical inaccuracies in that which cause issues for some edges with certain winding angles. That's my best guess anyway.

Comment (bugadani) [reply embedded-graphics/src/primitives/line_joint.rs:205]: Didn't I just rotate the triangle? I did not flip 2 points, but shifted everything.

**Comment by bugadani** [comment]
In this case, the overdraw is caused by the fact that we don't know any more, where the last edge segment started, so the intersection returns a point that is correct segment-wise, but not on the edge of the polilyne.

image

File: embedded-graphics/src/primitives/line_joint.rs
Comment (jamwaffles) [reply embedded-graphics/src/primitives/line_joint.rs:205]: Sorry yeah, I've done similar things in the past but swapping the last two points instead of rotating.

**Comment by jamwaffles** [comment]
IMO that overdraw is ok. I think it's pretty difficult to deal with all the self-intersecting cases, so that can be tackled in the future when we add transparency (or not, with a warning in the docs about it).

**Comment by bugadani** [comment]
Hmm I think my first screenshot has the same issue. With that, I think there should be no more (unexplained) overdraws

**Comment by bugadani** [comment]
This is a rollercoaster for sure.

image

But:
image

**Comment by bugadani** [comment]
Scanline iterators are hard.
image

But they are fast(er). Compared to the previous method (checking for each pixel if it's in the triangle), this one runs fast enough. On a 6 core ryzen 5...

I have no idea why the iterator is broken, though and I gave up for today.

**Comment by bugadani** [comment]
Current state of affairs:

I've implemented a scanline iterator that should not overdraw, but it can't differentiate edge points.
Results:
image
Almost as good as the triangle.contains() snail, but with a lot better perf and an added rendering artifact.

**Comment by jamwaffles** [comment]
So close! Is the only remaining issue the overdraw at sharp edges now?

> looking back more triangles

I haven't had time to look at your code in detail, but what do you mean by this? I'd be ok consuming a bit more memory if it means we get no overdraw in ordinary joints. Similarly to the thick lines feature, IMO the first step is to just get it working with a good unit test suite. We can optimise in the future, using the tests to ensure the same behaviour.

FWIW CI is failing on a formatting check.

**Comment by bugadani** [comment]
> I haven't had time to look at your code in detail

I plan on updating the PR description with what I've done so you can get an idea without looking at the code. Right now, I'm rendering the triangles using a new scanline iterator, and I keep track of the last rendered triangle to exclude some points - this works around the overdraw problem almost all the time.

Oh I don't think you'd want to waste 400 bytes of memory just to fix 2 pixels.

**Comment by jamwaffles** [comment]
> Oh I don't think you'd want to waste 400 bytes of memory just to fix 2 pixels.

That's... quite a lot of memory :(. Where does it all come from?

Overdraw will bite us in the ass when we add transparency though so maybe it's good to start with a completely non-overdrawing solution, regardless of memory usage.

**Comment by bugadani** [comment]
> Where does it all come from?

I kinda pulled that number out of thin air and is actually 3xOption<Triangle> which is hmm... 3 * (12 + 4) = 64, I think. But I also think I can do better :)

**Comment by jamwaffles** [comment]
I don't know if this would be useful to you or not, but I did try a trapezium iterator in https://github.com/jamwaffles/embedded-graphics/pull/386/commits/b5dbd99c9bffb8ffefacf9dd21be334b91bf7db0 which might save you some memory. It replaces the two triangles that make up the straight line segment with one shape. Might save some memory if you can shoehorn it into your approach.

**Comment by bugadani** [comment]
Maybe it could, I'll look at it.

Clippy is rather helpful, by the way:

```
warning: large size difference between variants
  --> embedded-graphics\src\primitives\polyline\styled.rs:17:5
   |
17 |     Thick(ThickPoints<'a>),
   |     ^^^^^^^^^^^^^^^^^^^^^^ this variant is 488 bytes
```

**Comment by bugadani** [comment]
```
warning: large size difference between variants
  --> embedded-graphics\src\primitives\polyline\styled.rs:17:5
   |
17 |     Thick(ThickPoints<'a>),
   |     ^^^^^^^^^^^^^^^^^^^^^^ this variant is 400 bytes
```

Getting better.

File: embedded-graphics/src/primitives/polyline/triangle_iterator.rs
Comment (jamwaffles) [inline embedded-graphics/src/primitives/polyline/triangle_iterator.rs:79]: Can you check if this is still valid?

File: embedded-graphics/src/primitives/polyline/triangle_iterator.rs
Comment (jamwaffles) [inline embedded-graphics/src/primitives/polyline/triangle_iterator.rs:97]: Likewise here

File: embedded-graphics/src/primitives/triangle/fill_scanline_iterator.rs
Comment (jamwaffles) [inline embedded-graphics/src/primitives/triangle/fill_scanline_iterator.rs:95]: Does `f` stand for `first` and `l` for `last` her? It would be better to use longer, more descriptive variable names IMO.

Can you also add a short doc comment description to this method explaining what it does? I think I understand, but a few words would help me and others.

File: embedded-graphics/src/primitives/triangle/mod.rs
Comment (jamwaffles) [inline embedded-graphics/src/primitives/triangle/mod.rs:124]: This can be removed now I think?

File: embedded-graphics/src/primitives/triangle/mod.rs
Comment (jamwaffles) [inline embedded-graphics/src/primitives/triangle/mod.rs:212]: This could be condensed using Rust's short circuiting rules:

```suggestion
        self.bounding_box().contains(*point) && in_mathematical_triangle(point)
```

**Review by jamwaffles — CHANGES_REQUESTED** [review]
Nice work on this! I found a few small things.

File: embedded-graphics/src/primitives/triangle/mod.rs
Comment (bugadani) [reply embedded-graphics/src/primitives/triangle/mod.rs:212]: I didn't want to go fancy :)

**Comment by jamwaffles** [comment]
Just to note that the `thick polyline` benchmark is 270% slower on my machine compared to `jamwaffles/line-joints` :grimacing: 

If the rendering is correct though, who am I to complain...!

**Comment by bugadani** [comment]
> Just to note that the `thick polyline` benchmark is 270% slower on my machine compared to `jamwaffles/line-joints` 😬
> 
> If the rendering is correct though, who am I to complain...!

Now with 3x as much checking per pixel!
Honestly, I'm surprised it's not worse.

File: embedded-graphics/src/primitives/triangle/fill_scanline_iterator.rs
Comment (bugadani) [reply embedded-graphics/src/primitives/triangle/fill_scanline_iterator.rs:128]: Yeah this is not nice :)

**Comment by bugadani** [comment]
@jamwaffles May or may not be my fault, but stroke alignments other than center aren't beveled correctly.
image

**Comment by jamwaffles** [comment]
That's a bug on my branch as well, whew. There's no overdraw though!

**Comment by bugadani** [comment]
@jamwaffles I've optimized Triangle::contains for the polylines - this cut rendering time in half. However, I think I saw some regression in the triangle benches. Should I move this contains function to a polyline-only place?

**Comment by bugadani** [comment]
Right now, this is much better than it was this morning.

```
Without lookback:
thick polyline          time:   [32.778 us 32.841 us 32.914 us]

With lookback:
thick polyline          time:   [90.987 us 91.333 us 91.737 us]

Original:
thick polyline          time:   [47.859 us 48.020 us 48.240 us]
```

File: embedded-graphics/src/primitives/polyline/thick_points.rs
Comment (bugadani) [inline embedded-graphics/src/primitives/polyline/thick_points.rs:59]: Technically speaking, one of this can be elided. When the joint turns right, prev_triangle2 check is redundant, and when it turns left, prev_triangle3 is. Also, probably both can be elided for mitered joints, but I haven't tried that.

prev_triangle is necessary though to avoid overdrawing the joining edges.

Comment (bugadani) [reply embedded-graphics/src/primitives/polyline/thick_points.rs:59]: Although I think it may be better to find a correct solution instead of hammering out the issues with the current naive one.

**Comment by bugadani** [comment]
Darn it, last commits have a regression... :'(

**Comment by bugadani** [comment]
image

**Comment by bugadani** [comment]
I guess (did not confirm but symptoms seem to indicate that) I found a case where Bresenham's algo fails to return a point that is otherwise inside a triangle.

**Comment by jamwaffles** [comment]
I think I've seen that before as well. The _center_ of a point can lie on either side of the true line when rendered using the Bresenham algorithm, up to 0.5px away IIRC. This might be what you're seeing.

**Comment by bugadani** [comment]
@jamwaffles there are 2 regression tests I had to #[ignore] because they require a MockDisplay > 64x64. Can we increase MD's size so these tests fit?

**Comment by jamwaffles** [comment]
We could, but I'd prefer to find a set of points that tests for the same regression, but inside the `MockDisplay` area.

**Comment by bugadani** [comment]
I really should set rustfmt as a precommit hook script.

@jamwaffles I think this is ready for you. I don't want to delay the feature with optimizations that may or may not work.

File: embedded-graphics/src/primitives/triangle/mod.rs
Comment (bugadani) [inline embedded-graphics/src/primitives/triangle/mod.rs:437]: @jamwaffles what's your issue here? Comment says "should also return true" and then you assert that it returns false.

**Review by jamwaffles — CHANGES_REQUESTED** [review]
Looks pretty good, but I'd like to see more doc/inline comments explaining what things to and how all the pieces fit together. This makes it easier for me to understand and pick up in my PR, and also for other maintainers to hack on.

File: embedded-graphics/src/primitives/triangle/mod.rs
Comment (bugadani) [reply embedded-graphics/src/primitives/triangle/mod.rs:97]: But even I don't know 🤣 j/k will do

**Review by jamwaffles — APPROVED** [review]
Looks good now. Thanks for progressing this feature! I was reaching my wits end tbh. In terms of attribution, would you prefer I rebase this onto my branch, or squash and merge it?

**Comment by bugadani** [comment]
> Looks good now. Thanks for progressing this feature! I was reaching my wits end tbh. In terms of attribution, would you prefer I rebase this onto my branch, or squash and merge it?

I'll leave that up to you. As far as I can see, the commit history is linear with your branch so both works.

To quote a colleague of mine: "The problem with github's squash and merge feature is that it's not a squash and it's not a merge" :)

I'm fine with a squash.

**Comment by jamwaffles** [comment]
Well, seeing as we're going for the record for most commits in a single PR I've rebased it ;)

**Pull request merged**
